File: frontend/src/api/screener.ts
/**
 * 个股挖掘 API 接口
 * 提供龙头高度、连板统计、涨跌停等数据
 */

import request from './request';
import { config } from '../config';

// ============================================
// 类型定义
// ============================================

export interface LeaderItem {
  code: string;
  name: string;
  sector: string;
  leader_height: number;
  sector_rank: number;
  total_stocks: number;
  price: number;
  change_percent: number;
  amount: number;
}

export interface ConsecutiveBoardItem {
  code: string;
  name: string;
  sector: string;
  consecutive_days: number;
  start_date: string;
  end_date: string;
  board_type: string;
  current_price: number;
  reason: string;
}

export interface LimitItem {
  code: string;
  name: string;
  sector: string;
  limit_type: string;
  limit_time: string;
  limit_price: number;
  volume: number;
  amount: number;
  reason: string;
  is_first: boolean;
}

// ============================================
// API 函数
// ============================================

/**
 * 获取龙头高度排行榜
 * @param sector 板块代码（可选）
 * @param limit 返回数量限制
 * @returns 龙头高度列表
 */
export async function fetchLeaders(
  sector?: string,
  limit: number = 50
): Promise<LeaderItem[]> {
  const params = new URLSearchParams();
  if (sector) params.append('sector', sector);
  params.append('limit', limit.toString());

  return request.get(`${config.storageUrl}/api/screener/leaders?${params.toString()}`);
}

/**
 * 获取连板统计数据
 * @param minDays 最小连板天数
 * @param boardType 连板类型："连涨" 或 "连跌"
 * @param limit 返回数量限制
 * @returns 连板统计列表
 */
export async function fetchConsecutiveBoards(
  minDays: number = 2,
  boardType: string = '连涨',
  limit: number = 50
): Promise<ConsecutiveBoardItem[]> {
  return request.get(
    `${config.storageUrl}/api/screener/consecutive?min_days=${minDays}&board_type=${boardType}&limit=${limit}`
  );
}

/**
 * 获取涨停股票列表
 * @param date 日期（"today" 或具体日期）
 * @param limit 返回数量限制
 * @returns 涨停股票列表
 */
export async function fetchLimitUp(
  date: string = 'today',
  limit: number = 50
): Promise<LimitItem[]> {
  return request.get(`${config.storageUrl}/api/screener/limit-up?date=${date}&limit=${limit}`);
}

/**
 * 获取跌停股票列表
 * @param date 日期（"today" 或具体日期）
 * @param limit 返回数量限制
 * @returns 跌停股票列表
 */
export async function fetchLimitDown(
  date: string = 'today',
  limit: number = 50
): Promise<LimitItem[]> {
  return request.get(`${config.storageUrl}/api/screener/limit-down?date=${date}&limit=${limit}`);
}

/**
 * 实时计算龙头高度
 * @param sectorCode 板块代码
 * @returns 龙头高度列表
 */
export async function fetchLeadersRealtime(
  sectorCode: string
): Promise<LeaderItem[]> {
  return request.get(`${config.storageUrl}/api/screener/leaders-realtime?sector_code=${sectorCode}`);
}

/**
 * 实时计算连板天数
 * @param code 股票代码
 * @returns 连板天数
 */
export async function fetchConsecutiveRealtime(code: string): Promise<number> {
  return request.get(`${config.storageUrl}/api/screener/consecutive-realtime?code=${code}`);
}

/**
 * 实时检测涨跌停
 * @returns 涨跌停股票列表
 */
export async function detectLimitStocks(): Promise<LimitItem[]> {
  return request.get(`${config.storageUrl}/api/screener/limit-detect`);
}

File: services/query-service/src/screener.rs
// 个股挖掘模块
//
// 功能：
// - 龙头高度计算
// - 连板天统计
// - 涨停跌停分析

use clickhouse::Client;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct LeaderItem {
    pub code: String,
    pub name: String,
    pub sector: String,
    pub leader_height: f64, // 龙头高度：行业排名倒数
    pub price: f64,
    pub change_percent: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ConsecutiveBoardItem {
    pub code: String,
    pub name: String,
    pub consecutive_days: i32,
    pub start_date: String,
    pub end_date: String,
    pub reason: String, // 涨停原因
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LimitItem {
    pub code: String,
    pub name: String,
    pub limit_time: String, // 涨停/跌停时间
    pub limit_price: f64,
    pub volume: f64,
    pub amount: f64,
    pub reason: String,
}

// ScreenerManager 核心管理类
pub struct ScreenerManager {
    client: Client,
}

impl ScreenerManager {
    pub fn new(client: Client) -> Self {
        Self { client }
    }

    // 计算龙头高度
    // 龙头高度 = 1 / 行业内排名
    pub async fn calculate_leader_height(
        &self,
        sector: &str,
    ) -> Result<Vec<LeaderItem>, anyhow::Error> {
        // TODO: 实现龙头高度计算逻辑
        // 1. 查询行业内所有股票
        // 2. 按市值或涨跌幅排序
        // 3. 计算龙头高度
        Ok(vec![])
    }

    // 统计连板天数
    pub async fn get_consecutive_boards(
        &self,
        min_days: i32,
    ) -> Result<Vec<ConsecutiveBoardItem>, anyhow::Error> {
        // TODO: 实现连板统计逻辑
        // 1. 查询历史涨停数据
        // 2. 计算连续涨停天数
        // 3. 返回连板股票列表
        Ok(vec![])
    }

    // 查询涨停股票
    pub async fn get_limit_up_stocks(&self, date: &str) -> Result<Vec<LimitItem>, anyhow::Error> {
        // TODO: 实现涨停查询逻辑
        // 1. 查询当日涨停股票
        // 2. 统计涨停时间和金额
        // 3. 分析涨停原因
        Ok(vec![])
    }

    // 查询跌停股票
    pub async fn get_limit_down_stocks(&self, date: &str) -> Result<Vec<LimitItem>, anyhow::Error> {
        // TODO: 实现跌停查询逻辑
        Ok(vec![])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_leader_height_calculation() {
        // TODO: 添加单元测试
    }

    #[test]
    fn test_consecutive_boards() {
        // TODO: 添加单元测试
    }

    #[test]
    fn test_limit_stocks() {
        // TODO: 添加单元测试
    }
}
